File: OS X/TPWindowManager.m
//
//  TPWindowManager.m
//  TwinPeaks
//
//  Created by marc hoffman on 10/1/11.
//  Copyright (c) 2011 __MyCompanyName__. All rights reserved.
//

#import "TPWindowManager.h"

@implementation TPWindowManager

static TPWindowManager *instance;

+ (TPWindowManager *)sharedInstance
{
    if (!instance) instance = [[self alloc] init];
    return instance;
}

- (NSWindowController *)uniqueWindowControllerForClass:(Class)class
{
    if (!singleWindows) singleWindows = [[NSMutableDictionary dictionaryWithCapacity:5] retain];
	NSWindowController *w = [singleWindows objectForKey:class];
	if (!w)
	{
		w = [[[class alloc] init] autorelease];
		[singleWindows setObject:w forKey:class];
	}
	return w;
}

- (void)showWindowForClass:(Class)class
{
	NSWindowController *w = [self uniqueWindowControllerForClass:class];
	[w showWindow:nil];
}

@end

File: OS X (C#)/TPWindowManager.cs
using AppKit;

namespace RemObjects.TwinPeaks.OSX
{
	public class TPWindowManager: NSObject
	{
		protected NSMutableDictionary singleWindows;

		private static TPWindowManager _sharedInstance;
		public static TPWindowManager sharedInstance()
		{
			if (_sharedInstance == null)
				_sharedInstance = new TPWindowManager();
			return _sharedInstance;
		}

		public NSWindowController uniqueWindowControllerForClass(Class @class)
		{
			if (!(singleWindows != null))
				singleWindows = NSMutableDictionary.dictionaryWithCapacity(5);
			NSWindowController w = singleWindows.objectForKey(@class);
			if (!(w != null))
			{
				w = @class.alloc().init();
				singleWindows.setObject(w) forKey(@class);
			}
			return w;
		}

		public void showWindowForClass(Class @class)
		{
			NSWindowController w = this.uniqueWindowControllerForClass(@class);
			w.showWindow(null);
		}
	}
}
